### DISCUSSION TITLE
How to create PyTorch Dataset and Dataloader from multiple point cloud files

### DISCUSSION THREAD

**Discussion by davodogster** [pyg-team/pytorch_geometric#4089] [Q&A] [answered]
Hi there, Awesome work on the repo!

I have individual point cloud files (e.g a "per object" .txt or .las file containing the X,Y,Z's and a separate file for the target) - It's a multitask problem where each point cloud has 2 associated labels.
Could you give me any pointers on how I can convert this data into a PyTorch dataset or use it in a Dataloader for 3D Deep Learning?
To make things more complicated the point clouds all contain a different number of points.

I have been searching issues and looking for information but haven't come across much for this. Every repo for images and point clouds I would have thought the most important thing would be "train on custom dataset" or "dataloader for custom dataset".
Cheers, Sam

**Comment by rusty1s** [comment, marked as answer]
For each point cloud, you would create an individual `data` object, e.g.:
```python
import pandas as pd

pos = torch.from_numpy(pd.read_csv(pos.txt).values).to(torch.float)
y = ... # read label into PyTorch tensor
data = Data(pos=pos, y=y)
```
You can then simply hold that in a Python list or merge it into a PyG `InMemoryDataset` and pass it to the `DataLoader`:
```python
from torch_geometric.loader import DataLoader

loader = DataLoader([data1, data2, data3, ...], batch_size=16, ...)
```

**Reply by davodogster** [reply]
Thanks so much!! and what about for creating the dataset and reading from disk into the DL model? and each with an uneven number of points?

**Reply by rusty1s** [reply]
For persisting the dataset on disk, you can apply the above scheme in the `process` method of an `InMemoryDataset`. Our tutorial goes over that in detail. Mini-batching will also work for a varying number of points.

**Reply by davodogster** [reply]
Hi @rusty1s I have loaded in my regression point cloud dataset:
![image](https://user-images.githubusercontent.com/46079516/157775436-c41400dc-1dfa-46e1-bc33-845b56f64d0d.png)

Trying to Regress the cylinder Diameter which I have ground truth for.

![image](https://user-images.githubusercontent.com/46079516/157775634-62c26f92-531a-4d47-bba0-881d1d6f67d6.png)

 and I have figured out how to train the PointNet with MessagePassing from here:
https://colab.research.google.com/drive/1D45E5bUK3gQ40YpZo65ozs7hg5l-eo_U?usp=sharing#scrollTo=TkV7DDBxLmUR

![image](https://user-images.githubusercontent.com/46079516/157775709-5a650566-83a2-4905-9367-28be63f8106f.png)

The model trains and learns (loss = MSE) but does not give desirable RMSE.
I tried training the model for 100 epochs but it didn't improve.

Are there any other networks that may give better results? Perhaps a network that captures global geometry? I'm still trying to figure out how to write a model class from scratch because I know some models use message passing, others don't, some models take 'norms' and 'batch' as input and others don't.. it's a lot to understand and learn but very interesting.
Thank you, Sam

**Reply by rusty1s** [reply]
I think it should still be possible for PointNet to learn this global property. I think you might have more success when normalizing both inputs (center point clouds, rotation aligned) and labels (scale them to a fixed interval).

**Reply by davodogster** [reply]
Thanks! Am currently training with these adjustments. Training MSE can get very good even without scaling the labels. What capability does PyG have for preventing overfitting, e.g TopKPooling? Remembering my Data is sparse / each PC is a different size. I wasn't able to get TopKPooling to work - it gave a tensor dimension error. Thanks, Sam

**Reply by rusty1s** [reply]
You might have some success in avoiding overfitting by making use of dropout (both on node features and edges via `torch.nn.functional.dropout` and `torch_geometric.utils.dropout_adj`).

**Reply by davodogster** [reply]
Hi @rusty1s , thanks I've been able to implement both of those dropout layers and they do give some improvement. I will warn that torch.nn.functional.dropout gives different predictions on different runs even when using model.eval() so I will be switching to torch.nn.Dropout. Also, for F.dropout If I add training = self.training it makes the model train very poorly train and val loss are much higher) compared to when I omit the training flag.

**Reply by rusty1s** [reply]
`training=self.training` should disable dropout during evaluation (`model.eval()`). As such, `F.dropout` and `torch.nn.Dropout` should behave equivalent.
